From 6e4185bf986106991e5ae22dee1989bebe804411 Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Thu, 31 Aug 2017 09:39:12 +0800
Subject: [PATCH 2469/5242] MMFMWK-7674: PxP: add YVU420P support

commit  13bff5f49cc2f329db06a21e74ce30839e9f42b8 from
https://source.codeaurora.org/external/imx/linux-imx.git

PxP PS engine support YUV420 format, but not YVU420. The difference
between two format is U and V, if we exchange U and V base address,
the PxP driver can also support YVU420 format.

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
Reviewed-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit cbc71da10afe67ae815a97a68e4a1b0dff4693eb)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/pxp/pxp_dma_v3.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/dma/pxp/pxp_dma_v3.c b/drivers/dma/pxp/pxp_dma_v3.c
index 9359440..c2a0133 100644
--- a/drivers/dma/pxp/pxp_dma_v3.c
+++ b/drivers/dma/pxp/pxp_dma_v3.c
@@ -1852,6 +1852,7 @@ static bool fmt_ps_support(uint32_t format)
 	case PXP_PIX_FMT_NV21:
 	case PXP_PIX_FMT_YUV422P:
 	case PXP_PIX_FMT_YUV420P:
+	case PXP_PIX_FMT_YVU420P:
 	case PXP_PIX_FMT_RGBA32:
 	case PXP_PIX_FMT_RGBX32:
 	case PXP_PIX_FMT_RGBA555:
@@ -2379,7 +2380,13 @@ static int pxp_ps_config(struct pxp_pixmap *input,
 			pxp_writel(U + (offset >> 2), HW_PXP_PS_UBUF);
 			V = U + (input->width * input->height >> 2);
 			pxp_writel(V + (offset >> 2), HW_PXP_PS_VBUF);
+		} else if (input->format == PXP_PIX_FMT_YVU420P) {
+			U = input->paddr + input->width * input->height;
+			V = U + (input->width * input->height >> 2);
+			pxp_writel(U + (offset >> 2), HW_PXP_PS_VBUF);
+			pxp_writel(V + (offset >> 2), HW_PXP_PS_UBUF);
 		}
+
 		break;
 	default:
 		break;
-- 
1.7.9.5

